### PR TITLE
[move] Stateful (de)serialization for function value lookups & delayed fields improvements

### PR DESCRIPTION
**Pull request by georgemitenkov**
## Description

This PR refactors value (de)serialization. Changes include:
- `DelayedFieldID` is used instead of generic types. We do not mock it in testing, and this is just a `u64`, so the complexity of associated types, generics and trait bounds is simply not worth it and makes the code more verbose. Also we had some patterns like `DelayedFieldID` to `u64` and then `u64` back to `DelayedFieldID` due to using generics.
- Changed (de)serialization to be optionally stateful. Users can add "extensions" that will do something during the process. Delayed fields is one of the options. Delayed fields changes are fully compatible.
- Changed V2 loader to use more functionality from `ModuleStorage`. This way, APIs are uniform for both MoveVM and Block-STM (which does use (de)serialization of Move values so we **need** this).
- Implemented a "function extension" - a way to lookup modules from the storage to get function arg types for function value serialization. See notes below.

Function value arg (de)serialization:
1.  When seeing a `ValueImpl::Function(...)` we can dispatch to the extension, like with delayed fields (see example in the code). We will get a vector of `Types` back.
2. I am relying on the assumption that function type args are stored as `TypeTag`s. It is a bit wasteful to construct runtime types and then layouts, so right now extension returns runtime types. They have all the info for (de)serialization. Note that ideally, we want to pass a slice of values + mask, so that we do not copy args, and instantiate only needed types. But this is enough for a prototype (similarly, we can lookup based on function index, and not a name).
3. Type info is stored alongside the function values. So in theory, we do not need to reconstruct types for serialization of values. But this depends on the interpreter implementation, as right now we have different stacks for values and types.

## How Has This Been Tested?

1. Existing tests cover delayed field (de)serialization.
2. New test for function extension (trait part, no serde as this requires a value representation).
3. Replay run: https://github.com/aptos-labs/aptos-core/actions/runs/12406719070

## Key Areas to Review

Ensure that the refactoring preserves the behaviour.

## Type of Change

- [x] Refactoring

## Which Components or Systems Does This Change Impact?

- [x] Move/Aptos Virtual Machine

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by georgemitenkov** [comment]
* **#15594** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15594?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15594?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by wrwg — APPROVED** [review]
Looks generally OK to me. Lets see whether this passes replay and then potential land it.

I think we really want to avoid branching here also because of incremental verification in production. If this change goes into the next devnet, it can be tested in the field independently of other upcoming changes. Otherwise we would need to verify a full branch with tons of changes at once in prod, we want to avoid this.

File: third_party/move/tools/move-unit-test/src/extensions.rs [hunk 1]
```diff
@@ -50,9 +51,13 @@ pub(crate) fn new_extensions<'a>() -> NativeContextExtensions<'a> {
 
 /// Print the change sets for available native context extensions.
 #[allow(unused)]
-pub(crate) fn print_change_sets<W: Write>(_w: &mut W, extensions: &mut NativeContextExtensions) {
+pub(crate) fn print_change_sets<W: Write>(
+    w: &mut W,
+    extensions: &mut NativeContextExtensions,
```
Comment (ziaptos) [inline third_party/move/tools/move-unit-test/src/extensions.rs:56]: nit: `native_extensions`

File: third_party/move/tools/move-unit-test/src/extensions.rs [hunk 2]
```diff
@@ -64,8 +69,14 @@ fn create_table_extension(extensions: &mut NativeContextExtensions) {
 }
 
 #[cfg(feature = "table-extension")]
-fn print_table_extension<W: Write>(w: &mut W, extensions: &mut NativeContextExtensions) {
-    let cs = extensions.remove::<NativeTableContext>().into_change_set();
+fn print_table_extension<W: Write>(
+    w: &mut W,
+    extensions: &mut NativeContextExtensions,
```
Comment (ziaptos) [inline third_party/move/tools/move-unit-test/src/extensions.rs:74]: same `native_extansions`

File: third_party/move/move-vm/runtime/src/storage/module_storage.rs [hunk 3]
```diff
@@ -302,3 +398,51 @@ where
...
+}
+
+impl<'a> FunctionValueExtension for FunctionExtensionAdapter<'a> {
+    fn get_function_arg_tys(
+        &self,
+        module_id: &ModuleId,
+        function_name: &IdentStr,
+        ty_arg_tags: Vec<TypeTag>,
+    ) -> PartialVMResult<Vec<Type>> {
+        let ty_args = ty_arg_tags
```
Comment (ziaptos) [inline third_party/move/move-vm/runtime/src/storage/module_storage.rs:427]: Maybe out of place comment, but we should start somewhere:

The ty_args, and ty names are used for different purposes. I'd have it:
`substitution_type_args` here, and in the lambda below `|ty| ....` -> `|type_to_substitute| ...`

File: third_party/move/move-vm/runtime/src/storage/module_storage.rs [hunk 4]
```diff
@@ -101,6 +113,90 @@ pub trait ModuleStorage: WithRuntimeEnvironment {
...
+                    st.name.as_ident_str(),
+                )
+                .map_err(|err| err.finish(Location::Undefined))
+            })
+            .map_err(|err| err.to_partial())
+    }
+
+    /// Returns the function definition corresponding to the specified name, as well as the module
+    /// where this function is defined. The returned module is also verified.
+    fn fetch_function_without_ty_args(
```
Comment (ziaptos) [inline third_party/move/move-vm/runtime/src/storage/module_storage.rs:178]: nit: why explicitly `without_ty_args`? in the name

File: third_party/move/extensions/move-table-extension/src/lib.rs [hunk 5]
```diff
@@ -685,13 +697,25 @@ fn get_table_handle(table: &StructRef) -> PartialVMResult<TableHandle> {
     Ok(TableHandle(handle))
 }
 
-fn serialize(layout: &MoveTypeLayout, val: &Value) -> PartialVMResult<Vec<u8>> {
-    val.simple_serialize(layout)
+fn serialize(
+    function_extension: &dyn FunctionValueExtension,
```
Comment (ziaptos) [inline third_party/move/extensions/move-table-extension/src/lib.rs:701]: nit: `function_value_extension` or short `fn_value_extension`.

Also, `FunctionValueExtension` can be short `FnValueExtension` but no strong opinions there

**Review by ziaptos — APPROVED** [review]
LGTM,

I'd just make sure to have more accurate namings:
`function_extansion` to `function_value_extension`..

File: third_party/move/move-vm/runtime/src/storage/module_storage.rs (hunk 4 above)
Comment (georgemitenkov) [reply third_party/move/move-vm/runtime/src/storage/module_storage.rs:178]: I think this should be called `fetch_function_definition` then? Basically types might be generic, this is not instantiated function yet, and to get it you do not provide type args.

File: third_party/move/extensions/move-table-extension/src/lib.rs (hunk 5 above)
Comment (georgemitenkov) [reply third_party/move/extensions/move-table-extension/src/lib.rs:701]: Renamed to `function_value_extension` everywhere

**Pull request merged**